tests/std/tests/P1132R7_out_ptr/test.cpp

```diff
@@ -0,0 +1,201 @@
+            static int i = 15;
+            *ptr         = &i;
+        };
+
+        f(inout_ptr(int_ptr));
+
+        assert(*int_ptr == 15);
+    }
+}
+
+void test_shared_ptr() {
+    shared_ptr<int> int_ptr = make_shared<int>(5);
+
+    {
+        const auto f = [](int** ptr) { *ptr = new int(7); };
+
+        f(out_ptr(int_ptr, default_delete<int>{}));
+
+        assert(*int_ptr == 7);
+    }
+    {
+        const auto f = [](void** ptr) { *ptr = new int(15); };
+
+        f(out_ptr(int_ptr, default_delete<int>{}));
+
+        assert(*int_ptr == 15);
+    }
+
+    int count          = 0;
+    const auto deleter = [&count](int* ptr) {
+        ++count;
+        delete ptr;
+    };
+
+    {
+        const auto f = [](int** ptr) { *ptr = new int(23); };
+
+        f(out_ptr(int_ptr, deleter));
+
+        assert(count == 0);
+        assert(*int_ptr == 23);
+    }
+    {
+        const auto f = [](void** ptr) { *ptr = new int(32); };
+
+        f(out_ptr(int_ptr, deleter));
+
+        assert(count == 1);
+        assert(*int_ptr == 32);
+    }
+
+    int_ptr.reset();
+    assert(count == 2);
+}
+
+template <class Ptr, class Pointer = void, class... Args>
+void test_smart_ptr(Args&&... args) {
+    Ptr int_ptr{new int(5)};
+
+    {
+        const auto f = [](int** ptr) {
+            assert(**ptr == 5);
+            *ptr = new int(7);
+        };
+
+        f(inout_ptr<Pointer>(int_ptr, forward<Args>(args)...));
+
+        assert(*int_ptr == 7);
+    }
+    {
+        const auto f = [](int** ptr) { *ptr = new int(12); };
+
+        f(out_ptr<Pointer>(int_ptr, forward<Args>(args)...));
+
+        assert(*int_ptr == 12);
+    }
+    {
+        const auto f = [](void** ptr) {
+            assert(*reinterpret_cast<int*>(*ptr) == 12);
+            *ptr = new int(15);
+        };
+
+        f(inout_ptr<int*>(int_ptr, forward<Args>(args)...));
+
+        assert(*int_ptr == 15);
+    }
+    {
+        const auto f = [](void** ptr) { *ptr = new int(19); };
+
+        f(out_ptr<int*>(int_ptr, forward<Args>(args)...));
+
+        assert(*int_ptr == 19);
+    }
+}
+
+struct reset_tag {};
+
+struct resettable_ptr {
+    using element_type = int; // test having element_type only
+
+    unique_ptr<int> ptr;
+
+    explicit resettable_ptr(int* p) : ptr(p) {}
+
+    void reset(int* p, reset_tag) {
+        ptr.reset(p);
+    }
+
+    auto operator*() const {
+        return *ptr;
+    }
+
+    auto get() const {
+        return ptr.get();
+    }
+
+    void release() {
+        ptr.release();
+    }
+};
+
+struct constructible_ptr {
+    using pointer = int*; // test having pointer only
+
+    unique_ptr<int> ptr;
+
+    explicit constructible_ptr(int* p) : ptr(p) {}
+    explicit constructible_ptr(int* p, reset_tag) : ptr(p) {}
+
+    auto operator*() const {
+        return *ptr;
+    }
+
+    auto get() const {
+        return ptr.get();
+    }
+
+    void release() {
+        ptr.release();
+    }
+};
+
+struct resettable_ptr2 : public resettable_ptr {

```

Comment:
We conventionally avoid saying `public` for `struct` inheritance (the rationale is that `struct` already strongly connotes that). Occurs once below, for `constructible_ptr2`.